br/pkg/stream/stream_metas.go

```diff
@@ -1082,23 +1223,135 @@ func (m MigrationExt) doTruncating(ctx context.Context, mig *pb.Migration, resul
 			m.tryRemovePrefix(ctx, compaction.GeneratedFiles, result)
 		}
 	}
-	for _, pfx := range mig.DestructPrefix {
-		m.tryRemovePrefix(ctx, pfx, result)
+}
+
+func (m MigrationExt) processExtFullBackup(ctx context.Context, mig *pb.Migration, result *MigratedTo) {
+	groups := LoadIngestedSSTss(ctx, m.s, mig.IngestedSstPaths)
+	processGroup := func(outErr error, e IngestedSSTss) (copyToNewMig bool, err error) {
+		if outErr != nil {
+			return true, outErr
+		}
+
+		if !e.GroupFinished() {
+			return true, nil
+		}
+
+		if e.GroupTS() >= mig.TruncatedTo {
+			return true, nil
+		}
+
+		for _, b := range e {
+			m.tryRemovePrefix(ctx, b.FilesPrefixHint, result)
+		}
+		return false, nil
 	}
+	for err, item := range iter.AsSeq(ctx, groups) {
+		copyToNewMig, err := processGroup(err, item)
+		if err != nil {
+			result.Warn(err)
+		}
+		if copyToNewMig {
+			for _, exb := range item {
+				result.NewBase.IngestedSstPaths = append(result.NewBase.IngestedSstPaths, exb.path)
+			}
+		}
+	}
+}
 
-	result.NewBase.TruncatedTo = mig.TruncatedTo
+type PathedIngestedSSTs struct {
+	*pb.IngestedSSTs
+	path string
+}
 
-	m.Hooks.StartLoadingMetaForTruncating()
-	mdSet := new(StreamMetadataSet)
-	mdSet.MetadataDownloadBatchSize = 128
-	shiftTS, err := mdSet.LoadUntilAndCalculateShiftTS(ctx, m.s, mig.TruncatedTo)
+type IngestedSSTss []PathedIngestedSSTs
+
+func (ebs IngestedSSTss) GroupFinished() bool {
+	for _, b := range ebs {
+		if b.Finished {
+			return true
+		}
+	}
+	return false
+}
+
+func (ebs IngestedSSTss) GroupTS() uint64 {
+	for _, b := range ebs {
+		if b.Finished {
+			return b.AsIfTs
+		}
+	}
+	return math.MaxUint64
+}
+
+func LoadIngestedSSTss(ctx context.Context, s storage.ExternalStorage, paths []string) iter.TryNextor[IngestedSSTss] {
+	fullBackupDirIter := iter.FromSlice(paths)
+	backups := iter.TryMap(fullBackupDirIter, func(name string) (PathedIngestedSSTs, error) {
+		// name is the absolute path in external storage.
+		bkup, err := readIngestedSSTs(ctx, name, s)
+		if err != nil {
+			return PathedIngestedSSTs{}, errors.Annotatef(err, "failed to read backup at %s", name)
+		}
+		return PathedIngestedSSTs{IngestedSSTs: bkup, path: name}, nil
+	})
+	extBackups, err := groupExtraBackups(ctx, backups)
 	if err != nil {
-		result.Warnings = append(result.Warnings, errors.Annotatef(err, "failed to open meta storage"))
-		return
+		return iter.Fail[IngestedSSTss](err)
 	}
-	m.Hooks.EndLoadingMetaForTruncating()
+	return iter.FromSlice(extBackups)
+}
 
-	m.doTruncateLogs(ctx, mdSet, shiftTS, result)
+func groupExtraBackups(ctx context.Context, i iter.TryNextor[PathedIngestedSSTs]) ([]IngestedSSTss, error) {
+	var (
+		collected = map[uuid.UUID]IngestedSSTss{}
+		finished  = map[uuid.UUID]struct{}{}
+	)
+
+	for {
+		res := i.TryNext(ctx)
+		if res.FinishedOrError() {
```

Comment:
let's say the file broken for some reasons. this will cause unmarshal error and stop collect and return partial sst files and no error. which seems not correct